x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Authenticator.java

```diff
@@ -0,0 +1,210 @@
+
+    /**
+     * Gets the token from the <code>Authorization</code> header if the header begins with
+     * <code>Bearer </code>
+     */
+    static SecureString extractBearerTokenFromHeader(ThreadContext threadContext) {
+        String header = threadContext.getHeader("Authorization");
+        if (Strings.hasText(header) && header.regionMatches(true,
+            0,
+            "Bearer ",
+            0,
+            "Bearer ".length()) && header.length() > "Bearer ".length()) {
+            char[] chars = new char[header.length() - "Bearer ".length()];
+            header.getChars("Bearer ".length(), header.length(), chars, 0);
+            return new SecureString(chars);
+        }
+        return null;
+    }
+
+    class Context implements Closeable {
+        private final ThreadContext threadContext;
+        private final AuthenticationService.AuditableRequest request;
+        private final User fallbackUser;
+        private final boolean allowAnonymous;
+        private final Realms realms;
+        private final List<AuthenticationToken> authenticationTokens = new ArrayList<>();
+        private final List<String> unsuccessfulMessages = new ArrayList<>();
+        private boolean handleNullToken = true;
+        private SecureString bearerString = null;
+        private List<Realm> defaultOrderedRealmList = null;
+        private List<Realm> unlicensedRealms = null;
+
+        public Context(
+            ThreadContext threadContext,
+            AuthenticationService.AuditableRequest request,
+            User fallbackUser,
+            boolean allowAnonymous,
+            Realms realms
+        ) {
+            this.threadContext = threadContext;
+            this.request = request;
+            this.fallbackUser = fallbackUser;
+            this.allowAnonymous = allowAnonymous;
+            this.realms = realms;
+        }
+
+        public ThreadContext getThreadContext() {
+            return threadContext;
+        }
+
+        public AuthenticationService.AuditableRequest getRequest() {
+            return request;
+        }
+
+        public User getFallbackUser() {
+            return fallbackUser;
+        }
+
+        public boolean isAllowAnonymous() {
+            return allowAnonymous;
+        }
+
+        public void setHandleNullToken(boolean value) {
+            handleNullToken = value;
+        }
+
+        public boolean shouldHandleNullToken() {
+            return handleNullToken;
+        }
+
+        public List<String> getUnsuccessfulMessages() {
+            return unsuccessfulMessages;
+        }
+
+        public void addAuthenticationToken(AuthenticationToken authenticationToken) {
+            authenticationTokens.add(authenticationToken);
+        }
+
+        public AuthenticationToken getAuthenticationToken() {
+            return authenticationTokens.isEmpty() ? null : authenticationTokens.get(authenticationTokens.size() - 1);
```

Comment:
I thought about having a similar method that returns the `AuthenticationToken` based on the requested class. But I chose to go with the current version mostly because of the Realms:

The realms are of its own chain and grouped into a single Authenticator. It is a bit awkward to for `RealmsAuthenticator` for the right `AuthenticationToken` since different realm can use different type of tokens. If we let `RealmsAuthenticator` delegate to each Realm to for this method, the problem is that the current `Realm` class does not have such a method. If we introduce the method for builtin realms, it may break custom realms out there. Since my intention for this refactoring is to be transparent, breaking custom realms felt unwanted. 

The current "get last authenticationToken` also helped reducing the complexity of refactoring because that is basically the logic of existing implementation. I don't think it is absolutely ideal, but it is one less thing to worry about and fewer files to touch for the current refactoring. My preference is that we could re-visit it when we decide to flatten the Realms sub-chain into the main chain.

Another less important thought that I had is that there might be multiple AuthenticationToken of the same type but extracted from different places, e.g. `UsernamePasswordToken` can come from either `Authorization` header, secondary header or request body. So we may need a bit more information than just the type for getting the right instance.

Overall, I'd prefer to keep the current version for this PR. I renamed it to `getMostRecentAuthenticationToken` and hopefully it makes the intention clear.